### PR TITLE
[forward port] Support DingTalk and Microsoft Teams as the notifier 

### PR DESCRIPTION
**Pull request by GGGitBoy**
Problem:

Microsoft Teams and DingTalk are all very popular IM tools. However, users cannot use them in Rancher out of box.

Solution:

AlertManager doesn't support them as a in-tree notifiers. 
We can use a generic `webhook` type alert in  AlertManager side and point it to a micro service for redirecting alert requests to Microsoft Teams and DingTalk.
The  micro service is responsible for changing the alert payload to fulfill MS Teams' API and redirecting the alert to MS Teams.
Also the micro service will be deployed on demand if there is no Microsoft Teams or DingTalk targets. It will not be deployed.
From Rancher users' perspective, they will be able to use Microsoft Teams and DingTalk out of box just like use other existing notifiers in Rancher UI such as Slack and email.

Related Issue: 

https://github.com/rancher/rancher/issues/15802
https://github.com/rancher/rancher/issues/19537

Related PR:

rancher/ui: https://github.com/rancher/ui/pull/3737 
rancher/system-charts: https://github.com/rancher/system-charts/pull/263
rancher/types: https://github.com/rancher/types/pull/1076

### CODE REVIEW

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 1]
```diff
@@ -35,6 +35,8 @@ var (
 	eventGroupInterval  = 1
 	eventGroupWait      = 1
 	eventRepeatInterval = 525600
+
+	systemProjectName = ""
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Do not declare the variable here, which is not threadsafe.

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 2]
```diff
@@ -542,6 +544,120 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
 
 				receiver.WebhookConfigs = append(receiver.WebhookConfigs, webhook)
 				receiverExist = true
+
+			} else if notifier.Spec.DingtalkConfig != nil {
+				appName, _ := monitorutil.ClusterAlertManagerInfo()
+				app, err := d.apps.GetNamespaced(systemProjectName, appName, metav1.GetOptions{})
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Use lister for Get

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 3]
```diff
@@ -542,6 +544,120 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
...
+				if err != nil {
+					errors.Wrapf(err, "get app %s failed", appName)
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Return early on errors. Otherwise, it is simply ignored.

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 4]
```diff
@@ -542,6 +544,120 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
...
+				}
+
+				if app.Spec.Answers["webhook-receiver.enabled"] != "true" {
+					app.Spec.Answers["webhook-receiver.enabled"] = "true"
+					d.apps.Update(app)
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Handle the error

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 5]
```diff
@@ -542,6 +544,120 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
...
+
+				webhookSecreteName, altermanagerAppNamespace := monitorutil.SecretWebhook()
+				secretClient := d.secretsGetter.Secrets(altermanagerAppNamespace)
+				configSecret, err := secretClient.Get(webhookSecreteName, metav1.GetOptions{})
+				if err != nil {
+					errors.Wrapf(err, "Get secrets")
+				}
+
+				oldConfig := configSecret.Data["config.yaml"]
+				config := make(map[string]map[string]map[string]string)
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Can we have a proper struct instead of handling a map like this?

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 6]
```diff
@@ -542,6 +544,120 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
...
+				config["receivers"] = make(map[string]map[string]string)
+				err = yaml.Unmarshal(oldConfig, &config)
+				if err != nil {
+					errors.Wrapf(err, "yaml Unmarshal config")
+				}
+
+				config["providers"]["dingtalk-"+r.NotifierName] = make(map[string]string)
+				config["receivers"][r.NotifierName] = make(map[string]string)
+				config["providers"]["dingtalk-"+r.NotifierName]["webhook_url"] = notifier.Spec.DingtalkConfig.URL
+				config["providers"]["dingtalk-"+r.NotifierName]["secret"] = notifier.Spec.DingtalkConfig.Secret
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: Use a variable/constant when a string shows up multiple times.

**Comment by GGGitBoy** [comment]
@gitlawr update

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 7]
```diff
@@ -542,6 +560,112 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
...
+					Provider: DingTalk + "-" + r.NotifierName,
+				}
+
+				data, err := yaml.Marshal(config)
+				if err != nil {
+					logrus.Errorf("Error yaml Marshal config : %v", err)
+				}
+
+				configSecret.Data["config.yaml"] = data
+				_, err = secretClient.Update(configSecret)
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: This code block looks duplicate with the next condition. Extract a standalone function

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 8]
```diff
@@ -71,6 +74,21 @@ type ConfigSyncer struct {
 	operatorCRDManager      *manager.PromOperatorCRDManager
 }
 
+type Config struct {
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: the name is too general

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 9]
```diff
@@ -542,6 +560,42 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
 
 				receiver.WebhookConfigs = append(receiver.WebhookConfigs, webhook)
 				receiverExist = true
+
+			} else if notifier.Spec.DingtalkConfig != nil {
+
+				err := d.syncReceiver(DingTalk, r.NotifierName, notifier.Spec.DingtalkConfig.URL, notifier.Spec.DingtalkConfig.Secret, notifier.Spec.DingtalkConfig.HTTPClientConfig)
+
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: clean up some of the new lines to make code style consistent

File: pkg/notifiers/sender.go [hunk 10]
```diff
@@ -73,6 +75,14 @@ func SendMessage(notifier *v3.Notifier, recipient string, msg *Message, dialer d
 		return TestWebhook(notifier.Spec.WebhookConfig.URL, msg.Content, notifier.Spec.WebhookConfig.HTTPClientConfig, dialer)
 	}
 
+	if notifier.Spec.DingtalkConfig != nil {
+		return TestDingtalk(notifier.Spec.DingtalkConfig.URL, notifier.Spec.DingtalkConfig.Secret, msg.Content, notifier.Spec.DingtalkConfig.HTTPClientConfig, dialer)
+	}
+
+	if notifier.Spec.MicrosoftConfig != nil {
```
Comment (gitlawr) [inline pkg/notifiers/sender.go]: Can we use MicrosoftTeams instead of Microsoft

File: tests/validation/tests/v3_api/test_alert.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    "type" : "/v3/schemas/microsoftConfig",
+    "url"  : "https://outlook.office.com/webhook",
+   
+}
+
+CLUSTER_ALERTING_APP = "cluster-alerting"
+ALERTMANAGER_CLUSTER_ALERTING_WORKLOAD  = "alertmanager-cluster-alerting"  	
+WEBHOOK_RECEIVER_WORKLOAD = "webhook-receiver"
+
+def test_add_notifier(remove_resource):
```
Comment (gitlawr) [inline tests/validation/tests/v3_api/test_alert.py]: Can we mock the services and test sending alerts

File: pkg/notifiers/sender.go [hunk 12]
```diff
@@ -241,6 +251,73 @@ func TestWebhook(url, msg string, cfg *v3.HTTPClientConfig, dialer dialer.Dialer
...
+	defer resp.Body.Close()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		return fmt.Errorf("HTTP status code is %d, not included in the 2xx success HTTP status codes", resp.StatusCode)
+	}
+
+	return nil
+}
+
+func TestMicrosoft(url, msg string, cfg *v3.HTTPClientConfig, dialer dialer.Dialer) error {
```
Comment (gitlawr) [inline pkg/notifiers/sender.go]: s/Microsoft/MicrosoftTeams/g

File: tests/validation/tests/v3_api/test_alert.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
+import pytest
+import requests
+
+from .common import *  # NOQA
+from .alert_common import MockReceiveAlert
+
+dingtalk_config = {
+    "type"  : "/v3/schemas/dingtalkConfig",
+    "url"   : "http://127.0.0.1:4050/dingtalk/",
```
Comment (gitlawr) [inline tests/validation/tests/v3_api/test_alert.py]: I wonder if the loopback address is accessible in the validation test environment. If that's the case, we can do it in integration tests for testing the send action. I will leave it to QA for verification

**Comment by gitlawr** [comment]
LGTM

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 14]
```diff
@@ -594,6 +647,13 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
 		}
 	}
 
+	if len(syncReceivers) > 0 {
+		err := d.syncReceiver(syncReceivers)
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go:651]: 1. It looks irrelevant to the purpose of `addRecipients()` which is to fill `alertconfig.Receiver` configurations.
2. It can be called multiple times given the way `addRecipients()` is invoked.

I think a proper place to do syncReceiver is in ConfigSyncer.sync() after alertmanager configs are handled.
Then get the latest receiver configs, update it only if it is changed.

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 15]
```diff
@@ -521,6 +545,28 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
 				receiver.WechatConfigs = append(receiver.WechatConfigs, wechat)
 				receiverExist = true
 
+			} else if notifier.Spec.DingtalkConfig != nil {
+
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go:549]: Nit: remove the new line

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 16]
```diff
@@ -521,6 +545,28 @@ func (d *ConfigSyncer) addRecipients(notifiers []*v3.Notifier, receiver *alertco
...
+				dingtalk := &alertconfig.WebhookConfig{
+					NotifierConfig: commonNotifierConfig,
+					URL:            webhookURL,
+				}
+
+				receiver.WebhookConfigs = append(receiver.WebhookConfigs, dingtalk)
+				receiverExist = true
+
+			} else if notifier.Spec.MSTeamsConfig != nil {
+
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go:560]: ditto

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 17]
```diff
@@ -658,3 +704,80 @@ func toAlertManagerURL(urlStr string) (*alertconfig.URL, error) {
...
+			}
+			if notifier.Spec.DingtalkConfig != nil {
+				provider := &Provider{
+					Type:       DingTalk,
+					WebHookURL: notifier.Spec.DingtalkConfig.URL,
+				}
+				if notifier.Spec.DingtalkConfig.Secret != "" {
+					provider.Secret = notifier.Spec.DingtalkConfig.Secret
+				}
+				if notifier.Spec.DingtalkConfig.HTTPClientConfig != nil {
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go:744]: Use `notifierutil.IsHTTPClientConfigSet()`

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 18]
```diff
@@ -658,3 +704,80 @@ func toAlertManagerURL(urlStr string) (*alertconfig.URL, error) {
...
+					Provider: r.NotifierName,
+				}
+				providers[r.NotifierName] = provider
+				receivers[r.NotifierName] = receiver
+			} else if notifier.Spec.MSTeamsConfig != nil {
+				provider := &Provider{
+					Type:       MicrosoftTeams,
+					WebHookURL: notifier.Spec.MSTeamsConfig.URL,
+				}
+				if notifier.Spec.MSTeamsConfig.HTTPClientConfig != nil {
```
Comment (gitlawr) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go:757]: Use notifierutil.IsHTTPClientConfigSet()

**Comment by GGGitBoy** [comment]
@gitlawr Update!

**Comment by guangbochen** [comment]
This is a forward port PR of 2.4 merged PR https://github.com/rancher/rancher/pull/27716.
